clients/drcachesim/tests/record_filter_unit_tests.cpp

```diff
@@ -427,6 +430,110 @@ test_encodings2regdeps_filter()
     return true;
 }
 
+/* Test SYS_futex preservation.
+ */
+#define SYS_futex_test 202
+#define SYS_fsync_test 74
+static bool
+test_func_marker_filter()
+{
+    constexpr addr_t syscall_base =
+        static_cast<addr_t>(func_trace_t::TRACE_FUNC_ID_SYSCALL_BASE);
+    constexpr addr_t syscall_num_futex = SYS_futex_test + syscall_base;
+    constexpr addr_t syscall_num_fsync = SYS_fsync_test + syscall_base;
+    constexpr addr_t PC = 0x7f6fdd3ec360;
+    constexpr addr_t ENCODING = 0xe78948;
+    std::vector<test_case_t> entries = {
+        /* Trace shard header.
+         */
+        { { TRACE_TYPE_HEADER, 0, { 0x1 } }, true, { true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VERSION, { 0x2 } }, true, { true } },
+        { { TRACE_TYPE_MARKER,
+            TRACE_MARKER_TYPE_FILETYPE,
+            { OFFLINE_FILE_TYPE_ARCH_X86_64 | OFFLINE_FILE_TYPE_ENCODINGS |
+              OFFLINE_FILE_TYPE_SYSCALL_NUMBERS | OFFLINE_FILE_TYPE_BLOCKING_SYSCALLS } },
+          true,
+          { true } },
+        { { TRACE_TYPE_THREAD, 0, { 0x4 } }, true, { true } },
+        { { TRACE_TYPE_PID, 0, { 0x5 } }, true, { true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, { 0x6 } },
+          true,
+          { true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT, { 0x3 } },
+          true,
+          { true } },
+
+        /* Chunk 1.
+         */
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0x7 } }, true, { true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0x8 } }, true, { true } },
+        /* We need at least one instruction with encodings.
```

Comment:
We do?

Reply:
If we only have function-related markers and no instruction, nothing gets output.
I did not investigate in detail why record_filter behaves that way, is it not supposed to happen? I can look into it.

Reply:
I think there are multiple cases like this due to the delayed state processing inherent in stream-oriented state machines.  Just augment the comment to make it clear *why* (meaning *who* needs this): as-is it sounds like a weird requirement of this func id filter itself.

Reply:
Got it, clarified that it's "record_filter" that needs it to output the trace.